### PR TITLE
coprocessor: add statistics back

### PR DESCRIPTION
**Pull request by choleraehyq**
@overvenus @siddontang @AndreMouche PTAL

Signed-off-by: Cholerae Hu <huyingqian@pingcap.com>

### CODE REVIEW

File: src/coprocessor/endpoint.rs [hunk 1]
```diff
@@ -486,9 +516,11 @@ pub fn check_if_outdated(deadline: Instant, tp: i64) -> Result<()> {
     Ok(())
 }
 
-fn respond(resp: Response, mut t: RequestTask) {
+fn respond(resp: Response, mut t: RequestTask) -> Statistics {
     t.stop_record_handling();
-    (t.on_resp)(resp)
+    let res = t.statistics.clone();
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: can we avoid clone here?

Comment (choleraehyq) [inline src/coprocessor/endpoint.rs]: I don't know :( @BusyJay

Comment (BusyJay) [inline src/coprocessor/endpoint.rs]: Of course.

File: src/coprocessor/endpoint.rs [hunk 2]
```diff
@@ -67,35 +67,70 @@ pub struct Host {
...
+                        .with_label_values(&[type_str, cf, tag])
+                        .observe(count as f64);
+                }
+            }
+        }
+    }
 }
 
-struct DummyContextFactory {}
+struct CopContextFactory {}
```
Comment (overvenus) [inline src/coprocessor/endpoint.rs]: Just `struct CopContextFactory;`.

File: src/storage/engine/mod.rs [hunk 3]
```diff
@@ -176,7 +176,7 @@ pub enum ScanMode {
 }
 
 /// Statistics collects the ops taken when fetching data.
-#[derive(Default)]
+#[derive(Default, Clone)]
```
Comment (overvenus) [inline src/storage/engine/mod.rs]: How about `Copy`?

File: src/storage/engine/mod.rs [hunk 4]
```diff
@@ -188,6 +188,8 @@ pub struct CFStatistics {
     pub seek_for_prev: usize,
 }
 
+unsafe impl Send for CFStatistics {}
```
Comment (overvenus) [inline src/storage/engine/mod.rs]: Why unsafe impl `Send`?

File: src/util/threadpool.rs [hunk 5]
```diff
@@ -20,8 +20,8 @@ use std::collections::VecDeque;
 use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering as AtomicOrdering};
 use std::fmt::Write;
 
-pub const DEFAULT_TASKS_PER_TICK: usize = 50;
-const WORKER_WAIT_TIME: u64 = 20; // ms
+pub const DEFAULT_TASKS_PER_TICK: usize = 500;
+const WORKER_WAIT_TIME: u64 = 500; // ms
```
Comment (siddontang) [inline src/util/threadpool.rs]: use 1s

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987 @BusyJay

**Comment by siddontang** [comment]
please fix the conflict.

PTAL @zhangjinpeng1987 @BusyJay

File: src/coprocessor/endpoint.rs [hunk 6]
```diff
@@ -63,35 +63,74 @@ pub struct Host {
...
-    fn on_tick(&mut self) {}
+    fn on_tick(&mut self) {
+        let task_count = self.task_count;
+        for type_str in &[STR_REQ_TYPE_SELECT, STR_REQ_TYPE_INDEX, STR_REQ_TYPE_DAG] {
+            let this_statistics = self.get_statistics(type_str);
+            for (cf, details) in this_statistics.details() {
+                for (tag, count) in details {
+                    COPR_SCAN_DETAILS
+                        .with_label_values(&[type_str, cf, tag])
+                        .observe(count as f64 / task_count as f64);
```
Comment (zhangjinpeng1987) [inline src/coprocessor/endpoint.rs]: How about task_count is 0.

**Pull request merged**